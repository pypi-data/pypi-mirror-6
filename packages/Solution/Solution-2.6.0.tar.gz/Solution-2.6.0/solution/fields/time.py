# -*- coding: utf-8 -*-
import datetime

from .. import validators as v
from ..utils import Markup, get_html_attrs
from .field import ValidationError
from .text import Text


class Time(Text):

    """A time field formatted as 'HH:mm AA'. Examples: 5:03 AM, 11:00 PM
    This field DOES NOT make any timezone conversions

    :param validate:
        An list of validators. This will evaluate the current `value` when
        the method `validate` is called.

    :param default:
        Default value. It must be a `datetime.time`.

    :param prepare:
        An optional function that takes the current value as a string
        and preprocess it before rendering.

    :param clean:
        An optional function that takes the value already converted to
        python and return a 'cleaned' version of it. If the value can't be
        cleaned `None` must be returned instead.

    :param hide_value:
        Do not render the current value a a string.

    """
    _type = 'time'
    default_validator = v.IsTime

    def __init__(self, **kwargs):
        kwargs.setdefault('default', None)
        return super(Time, self).__init__(**kwargs)

    def py_to_str(self, format=None, locale=None, **kwargs):
        tt = self.obj_value or self.default
        if not tt:
            return u''
        return tt.strftime('%l:%M %p').strip()

    def as_input(self, format=None, locale=None, **kwargs):
        kwargs['type'] = kwargs.setdefault('type', self._type)
        kwargs['name'] = self.name
        kwargs['value'] = self.to_string(**kwargs)
        if not self.optional:
            kwargs.setdefault('required', True)
        html = u'<input %s>' % get_html_attrs(kwargs)
        return Markup(html)

    def as_textarea(self, format=None, locale=None, **kwargs):
        kwargs['name'] = self.name
        if not self.optional:
            kwargs.setdefault('required', True)
        html_attrs = get_html_attrs(kwargs)
        value = self.to_string(**kwargs)
        html = u'<textarea %s>%s</textarea>' % (html_attrs, value)
        return Markup(html)

    def str_to_py(self, format=None, locale=None):
        if not self.str_value:
            return self.default or None
        try:
            num, p = self.str_value.upper().split(' ')
            splitted = [int(n) for n in num.split(':')]
            splitted.append(0)
            hour = splitted[0]
            minute = splitted[1]
            second = splitted[2]
            if p == 'PM':
                hour = hour + 12
            return datetime.time(hour, minute, second)
        except ValueError:
            raise ValidationError
