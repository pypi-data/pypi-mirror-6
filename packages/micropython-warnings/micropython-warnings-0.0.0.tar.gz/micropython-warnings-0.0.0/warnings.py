# MicroPython doesn't support warnings
