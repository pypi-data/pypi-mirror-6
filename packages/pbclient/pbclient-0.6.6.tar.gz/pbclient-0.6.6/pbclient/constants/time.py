DATE_FORMAT = '%Y-%m-%d'
DATE_FORMAT_HUMAN = 'YYYY-MM-DD'

