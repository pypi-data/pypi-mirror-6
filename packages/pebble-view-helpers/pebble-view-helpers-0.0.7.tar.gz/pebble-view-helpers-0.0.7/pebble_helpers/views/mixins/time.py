"""Mixins for class-based views to assist in managing time.
"""
__author__ = 'Scott Walton <scw@talktopebble.co.uk>'

from datetime import datetime

from dateutil import parser

from django.conf import settings


class CurrentDateTimeMixin(object):
    """Freezes the current datetime to the time at the start of the request.
    Helps with potential issues of a request crossing the day boundary.
    Also adds a TEST_DATETIME request argument that sets the datetime to a
    specific point during the testing environment.
    """
    def dispatch(self, request, *args, **kwargs):
        """Set the datetime for the request.
        """
        self.set_datetime(request)
        return super(
            CurrentDateTimeMixin, self).dispatch(request, *args, **kwargs)

    def set_datetime(self, request):
        """Freezes the datetime of the request. If TESTING is True, also reads
        the request arguments for a TEST_DATETIME that allows us to freeze
        the datetime during a testing environment.
        """
        request_args = getattr(request, request.method, {})
        DATETIME = 'TEST_DATETIME'

        if DATETIME in request_args and settings.TESTING:
            self._datetime = parser.parse(request_args[DATETIME])
        else:
            self._datetime = datetime.now()

    def get_date(self):
        """Get the date of the request starting.
        """
        return self._datetime.date()

    def get_datetime(self):
        """Get the full datetime of the request starting.
        """
        return self._datetime
