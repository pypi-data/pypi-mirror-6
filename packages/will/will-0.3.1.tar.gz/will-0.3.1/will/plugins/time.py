import datetime
from will.plugin import WillPlugin
from will.decorators import respond_to, periodic, hear, randomly, route, rendered_template


class TimePlugin(WillPlugin):

    @respond_to("what time is it")
    def what_time_is_it(self, message):
        now = datetime.datetime.now()
        self.say("It's %s." % self.to_natural_day_and_time(now), message=message)
