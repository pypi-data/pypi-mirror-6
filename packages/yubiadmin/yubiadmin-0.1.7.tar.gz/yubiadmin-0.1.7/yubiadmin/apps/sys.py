# Copyright (c) 2013 Yubico AB
# All rights reserved.
#
#   Redistribution and use in source and binary forms, with or
#   without modification, are permitted provided that the following
#   conditions are met:
#
#    1. Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#    2. Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import os
import subprocess
from webob import Response
from threading import Timer
from yubiadmin.util.app import App, render
from yubiadmin.util.system import run
from yubiadmin.apps.dashboard import panel

__all__ = [
    'app'
]


UPGRADE_LOG = "/var/tmp/yubix-upgrade"


def get_updates():
    s, o = run("apt-get upgrade -s | awk -F'[][() ]+' '/^Inst/{print $2}'")
    packages = o.splitlines()
    return packages


def needs_restart():
    return os.path.isfile('/var/run/reboot-required')


def reboot():
    run('reboot')


class Updater(object):
    def __init__(self):
        self.proc = subprocess.Popen('DEBIAN_FRONTEND=noninteractive '
                                     'apt-get -y dist-upgrade -o '
                                     'Dpkg::Options::="--force-confdef" -o '
                                     'Dpkg::Options::="--force-confold" | '
                                     'tee %s' % UPGRADE_LOG,
                                     stdout=subprocess.PIPE, shell=True)

    def __iter__(self):
        yield """
        <script type="text/javascript">
        function reload() {
            window.location.replace('/sys');
        }
        window.onload = function() {
            setTimeout(reload, 10000);
        }
        </script>
        <strong>Performing update, this may take a while...</strong><br/>
        <pre>
        """

        while True:
            line = self.proc.stdout.readline()
            if line:
                yield line
            else:
                yield '</pre><br /><strong>Update complete!</strong>'
                yield '<script type="text/javascript">reload();</script>'
                break


class SystemApp(App):
    """
    YubiX System
    """
    sections = ['general']
    priority = 30

    @property
    def disabled(self):
        #return not os.path.isdir('/usr/share/yubix')
        return False

    @property
    def hidden(self):
        return self.disabled

    @property
    def dash_panels(self):
        if needs_restart():
            yield panel('System', 'System restart required', level='danger')

        updates = len(get_updates())
        if updates > 0:
            yield panel(
                'System',
                'There are <strong>%d</strong> updates available' % updates,
                '/%s/general' % self.name,
                'info'
            )

        _, time = run('date "+%a, %d %b %Y %H:%M"')
        _, result = run('uptime')
        rest = [x.strip() for x in result.split('up', 1)][1]
        parts = [x.strip() for x in rest.split(',')]
        uptime = parts[0] if not 'days' in parts[0] else '%s, %s' % \
            tuple(parts[:2])
        yield panel('System', 'Date: %s<br />Uptime: %s' %
                   (time, uptime), level='info')

    def general(self, request):
        alerts = []
        if needs_restart():
            alerts.append({'message': 'The machine needs to reboot.',
                           'type': 'error'})
        return render('/sys/general', alerts=alerts, updates=get_updates())

    def update(self, request):
        run('apt-get update')
        return self.redirect('/sys')

    def dist_upgrade(self, request):
        if get_updates():
            return Response(app_iter=Updater())
        else:
            alerts = [{'message': 'Software is up to date!'}]
            return render('/sys/general', alerts=alerts)

    def reboot(self, request):
        if 'now' in request.params:
            run('reboot')
        else:
            timer = Timer(1, run, args=('reboot',))
            timer.start()
        alerts = [{'type': 'warn', 'message': 'Rebooting System...'}]
        return render('/sys/general', alerts=alerts)


app = SystemApp()
