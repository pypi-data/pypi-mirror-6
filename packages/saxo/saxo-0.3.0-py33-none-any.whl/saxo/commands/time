#!/usr/bin/env python3

# Copyright 2012-4, Sean B. Palmer
# Source: http://inamidst.com/saxo/

import time
import saxo

@saxo.pipe
def t(arg):
    return time.ctime()
