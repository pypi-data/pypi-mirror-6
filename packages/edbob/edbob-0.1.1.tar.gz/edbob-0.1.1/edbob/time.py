#!/usr/bin/env python
# -*- coding: utf-8  -*-
################################################################################
#
#  edbob -- Pythonic Software Framework
#  Copyright Â© 2010-2012 Lance Edgar
#
#  This file is part of edbob.
#
#  edbob is free software: you can redistribute it and/or modify it under the
#  terms of the GNU Affero General Public License as published by the Free
#  Software Foundation, either version 3 of the License, or (at your option)
#  any later version.
#
#  edbob is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
#  FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
#  more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with edbob.  If not, see <http://www.gnu.org/licenses/>.
#
################################################################################

"""
``edbob.time`` -- Date & Time Utilities
"""

import datetime
import pytz
import logging
import warnings

import edbob


__all__ = ['utc_time', 'local_time']

timezones = {}

log = logging.getLogger(__name__)


def init(config):
    """
    Reads configuration to become aware of all timezones which may concern the
    application.

    .. highlight:: ini

    The bare minimum configuration required is the ``zone.local`` setting::

       [edbob.time]
       zone.local = US/Pacific

    Multiple timezones may be configured like so::

       [edbob.time]
       zone.local = America/Los_Angeles
       zone.head_office = America/New_York
       zone.that_other_place = Asia/Manila
       zone.some_app = US/Central

    See `Wikipedia
    <http://en.wikipedia.org/wiki/List_of_tz_database_time_zones>`_ for a
    (presumably) full list of valid timezone names.

    .. note::
       A ``zone.utc`` option is automatically created for you; there is no need
       to define it.
    """

    for key in config.options('edbob.time'):
        if key.startswith('zone.'):
            tz = config.get('edbob.time', key)
            if tz:
                key = key[5:]
                log.debug("'%s' timezone set to '%s'" % (key, tz))
                set_timezone(tz, key)

    if 'local' not in timezones:
        tz = config.get('edbob.time', 'timezone')
        if tz:
            warnings.warn("Config option 'timezone' in 'edbob.time' section is deprecated.  "
                          "Please set 'zone.local' instead.",
                          DeprecationWarning)
            set_timezone(tz)
        else:
            log.warning("'local' timezone not configured; falling back to 'America/Chicago'")
            set_timezone('America/Chicago')

    set_timezone('UTC', 'utc')


def get_timezone(key='local'):
    """
    Returns the timezone referenced by ``key``.
    """

    if key not in timezones:
        edbob.config.require('edbob.time', 'zone.%s' % key)
    return timezones[key]


def set_timezone(tz, key='local'):
    """
    Stores a timezone in the global dictionary, using ``key``.

    ``tz`` must be a valid "Olson" time zone name, e.g. ``'US/Central'``.  See
    `Wikipedia <http://en.wikipedia.org/wiki/List_of_tz_database_time_zones>`_
    for a (presumably) full list of valid names.
    """

    timezones[key] = pytz.timezone(tz)


def local_time(stamp=None, from_='local', naive=False):
    """
    Returns a :class:`datetime.datetime` instance, with its ``tzinfo`` member
    set to the timezone referenced by the key ``'local'``.  If ``naive`` is
    ``True``, the result is stripped of its ``tzinfo`` member.

    If ``stamp`` is provided, and it is not already "aware," then its value is
    interpreted as being local to the timezone referenced by ``from_``.

    If ``stamp`` is not provided, the current time is assumed.
    """

    if not stamp:
        stamp = utc_time()
    elif not stamp.tzinfo:
        stamp = localize(stamp, from_=from_)
    return localize(stamp, naive=naive)


def localize(stamp, from_='local', to='local', naive=False):
    """
    Creates a "localized" version of ``stamp`` and returns it.

    ``stamp`` must be a :class:`datetime.datetime` instance.  If it is naive,
    its value is interpreted as being local to the timezone referenced by
    ``from_``.  If it is already aware (i.e. not naive), then ``from_`` is
    ignored.

    The timezone referenced by ``to`` is used to determine the final, "local"
    value for the timestamp.  If ``naive`` is ``True``, the timestamp is
    stripped of its ``tzinfo`` member before being returned.  Otherwise, it
    will remain aware of its timezone.
    """

    if not stamp.tzinfo:
        tz = get_timezone(from_)
        stamp = tz.localize(stamp)
    tz = get_timezone(to)
    stamp = stamp.astimezone(tz)
    if naive:
        stamp = stamp.replace(tzinfo=None)
    return stamp


def utc_time(stamp=None, from_='local', naive=False):
    """
    Returns a :class:`datetime.datetime` instance, with its ``tzinfo`` member
    set to the UTC timezone.  If ``naive`` is ``True``, the result is stripped
    of its ``tzinfo`` member.

    If ``stamp`` is provided, and it is not already "aware," then its value is
    interpreted as being local to the timezone referenced by ``from_``.

    If ``stamp`` is not provided, the current time is assumed.
    """

    if not stamp:
        stamp = datetime.datetime.utcnow()
        stamp = pytz.utc.localize(stamp)
    elif not stamp.tzinfo:
        stamp = localize(stamp, from_=from_)
    return localize(stamp, to='utc', naive=naive)
