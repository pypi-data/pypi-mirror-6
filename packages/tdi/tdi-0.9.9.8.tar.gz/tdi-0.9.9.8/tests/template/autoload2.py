#!/usr/bin/env python
import warnings as _warnings
_warnings.resetwarnings()
_warnings.filterwarnings('error')

import sys
import tempfile
import time

from tdi import html

html = html.replace(autoupdate=True)

tfile1 = tempfile.NamedTemporaryFile()
try:
    tfile1.write(
        """<html><body tdi="body" tdi:overlay=">foo">Yey</body></html>"""
    )
    tfile1.flush()

    tfile2 = tempfile.NamedTemporaryFile()
    try:
        tfile2.write("""<tdi tdi:overlay="foo">blub</tdi>""")
        tfile2.flush()

        # 2) Load the template from_file
        template = (
            html.from_file(tfile1.name)
            .overlay(html.from_file(tfile2.name))
        )
        print template.tree

        # (... wait for low-timer-resolution systems ...)
        time.sleep(3)

        # 3) Update the file
        tfile1.seek(0)
        tfile1.truncate()
        tfile1.write(
            """<xhtml><body tdi="nobody" tdi:overlay="foo">"""
            """Yup!</body></xhtml>"""
        )
        tfile1.flush()

        # 4) Voila
        print template.tree
    finally:
        tfile2.close()
finally:
    tfile1.close()
