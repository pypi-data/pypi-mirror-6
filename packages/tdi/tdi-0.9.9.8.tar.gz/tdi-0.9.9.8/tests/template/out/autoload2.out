/
  '<html>'
  body (<<< foo)
    'blub'
  '</html>'
\

/
  '<xhtml>'
  nobody (<<< foo)
    'blub'
  '</xhtml>'
\

